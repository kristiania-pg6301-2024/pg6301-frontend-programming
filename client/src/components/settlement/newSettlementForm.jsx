import React, {useState} from 'react';
import {BillDenominationTypes, CoinDenominationTypes} from "./domain";
import {BalanceInputForm} from "./balanceInputForm";

function sumBalance(balance) {
    let sum = 0;
    for (const denonimation in BillDenominationTypes) {
        if (balance[denonimation]?.count) {
            sum += balance[denonimation].count * BillDenominationTypes[denonimation].value;
        }
    }
    for (const denonimation in CoinDenominationTypes) {
        const type = CoinDenominationTypes[denonimation];
        if (balance[denonimation]?.count) {
            sum += balance[denonimation].count * type.value;
        } else if (balance[denonimation]?.weight) {
            sum += Math.round(balance[denonimation].weight / type.gramsPerCoin) * type.value;
        }
    }
    return sum;
}

export function NewSettlementForm() {
    const allDepartments = [
        "Auksjon", "Bilder", "Bøker", "Cafeteria", "Elektro", "Godteri", "Kjøkken", "Møbler", "Popcorn", "Sport",
    ];

    const [settlement, setSettlement] = useState({
        balance: {}
    })
    const sum = sumBalance(settlement.balance);

    return <form>
        <h2>Registrere kontanter</h2>
        <div>
            <div>Avdeling:</div>
            <select onChange={e => setSettlement(old => ({
                ...old,
                department: e.target.value
            }))}>
                <option></option>
                {allDepartments.map(d => <option key={d} value={d}>{d}</option>)}
            </select>
        </div>
        <BalanceInputForm
            balance={settlement.balance}
            setBalance={fn => setSettlement(old => ({
                ...old,
                balance: fn(old.balance)
            }))}
        />
        <div>
            <strong>Sum: {sum}</strong>
        </div>

        <pre>
            {JSON.stringify(settlement, null, 2)}
        </pre>
    </form>;
}
