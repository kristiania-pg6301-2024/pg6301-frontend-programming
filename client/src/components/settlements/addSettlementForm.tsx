import React from "react";

export function AddSettlementForm() {
  return (
    <form>
      <h1>Add settlement</h1>
      <button>Submit</button>
    </form>
  );
}
