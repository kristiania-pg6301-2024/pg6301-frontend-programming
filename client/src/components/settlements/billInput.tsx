import { BillType } from "./domain";
import React from "react";

interface Props {
  type: BillType;
  value?: number | undefined;
  setValue: (v: number) => void;
}

export function BillInput({ type: { key }, value, setValue }: Props) {
  return (
    <div key={key}>
      <label htmlFor={`newSettlementBalance_${key}`}>{key}:</label>
      <input
        id={`newSettlementBalance_${key}`}
        type={"number"}
        value={value || "0"}
        onChange={(e) => setValue(parseInt(e.target.value))}
      />
    </div>
  );
}
