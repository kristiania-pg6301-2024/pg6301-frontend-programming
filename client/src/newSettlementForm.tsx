import React from "react";

export function NewSettlementForm() {
  const departments = ["Møbler", "Elektronikk", "Bøker"];
  return (
    <>
      <h1>Register new settlement</h1>
      <div>
        <label>
          Avdeling:
          <select>
            {departments.map((department) => (
              <option key={department}>{department}</option>
            ))}
          </select>
        </label>
      </div>
    </>
  );
}
