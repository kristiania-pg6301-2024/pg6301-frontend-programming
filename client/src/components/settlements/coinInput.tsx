import { CoinAmount, CoinType } from "../../../../src/cashBalance";
import React, { Dispatch, SetStateAction } from "react";

interface Props {
  type: CoinType;
  value?: CoinAmount;
  setValue: Dispatch<SetStateAction<CoinAmount>>;
}

export function CoinInput({
  type: { key, gramsPerCoin },
  value,
  setValue,
}: Props) {
  return (
    <div key={key} className={`denomination${key}`}>
      <label>
        {key}:
        <input
          type={"number"}
          className={"count"}
          disabled={value && "grams" in value}
          value={
            !value
              ? ""
              : "count" in value
                ? value.count
                : Math.round(value.grams / gramsPerCoin)
          }
          onChange={(e) => setValue({ count: parseInt(e.target.value) })}
        />
        <input
          type={"number"}
          className={"grams"}
          disabled={value && "count" in value}
          value={
            !value
              ? ""
              : "grams" in value
                ? value.grams
                : value.count * gramsPerCoin
          }
          onChange={(e) => setValue({ grams: parseFloat(e.target.value) })}
        />
      </label>
    </div>
  );
}
