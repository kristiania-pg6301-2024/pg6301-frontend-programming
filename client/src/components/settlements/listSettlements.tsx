import { Settlement, sumBalance } from "../../../../src/cashBalance";
import { Link } from "react-router-dom";
import React from "react";

interface Props {
  settlements: Settlement[];
}

export function ListSettlements({ settlements }: Props) {
  return (
    <>
      <h1>Settlements</h1>
      {settlements.map(({ _id, department, balance }) => (
        <div key={_id}>
          {department}: {sumBalance(balance)} kr
        </div>
      ))}
      <Link to={"/new"}>New Settlement</Link>
    </>
  );
}
