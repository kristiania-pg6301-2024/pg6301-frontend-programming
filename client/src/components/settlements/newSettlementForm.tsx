import React, { useState } from "react";
import { billTypes, CashBalance, coinTypes } from "../../../../src/cashBalance";

const departments = ["Furniture", "Books"];

export interface Settlement {
  department: string;
  balance: CashBalance;
}

interface Props {
  onNewSettlement(settlement: Settlement): void;
}

export function NewSettlementForm({ onNewSettlement }: Props) {
  const [department, setDepartment] = useState("");
  const [balance, setBalance] = useState<CashBalance>({});
  function handleSubmit(e: React.FormEvent) {
    e.preventDefault();
    onNewSettlement({ department, balance });
  }

  function hasBalance() {
    return Object.values(balance).some((b) => !!b);
  }

  return (
    <form onSubmit={handleSubmit}>
      <h1>New Settlement</h1>

      <div>
        <label htmlFor={"newSettlementDepartment"}>Department: </label>
        <select
          id={"newSettlementDepartment"}
          value={department}
          onChange={(e) => setDepartment(e.target.value)}
        >
          <option></option>
          {departments.map((department) => (
            <option key={department}>{department}</option>
          ))}
        </select>
      </div>

      {billTypes.map(({ key }) => (
        <div key={key}>
          <label htmlFor={`newSettlementBalance_${key}`}>{key}:</label>
          <input
            id={`newSettlementBalance_${key}`}
            type={"number"}
            value={balance[key] || "0"}
            onChange={(e) =>
              setBalance((old) => ({ ...old, [key]: parseInt(e.target.value) }))
            }
          />
        </div>
      ))}
      <div className={"coins"}>
        {coinTypes.map(({ key, gramsPerCoin }) => (
          <div key={key} className={`denomination${key}`}>
            <label>
              {key}:
              <input
                type={"number"}
                className={"count"}
                disabled={balance[key] && "grams" in balance[key]}
                value={
                  !balance[key]
                    ? ""
                    : "count" in balance[key]
                      ? balance[key].count
                      : Math.round(balance[key].grams / gramsPerCoin)
                }
                onChange={(e) =>
                  setBalance((old) => ({
                    ...old,
                    [key]: { count: parseInt(e.target.value) },
                  }))
                }
              />
              <input
                type={"number"}
                className={"grams"}
                disabled={balance[key] && "count" in balance[key]}
                value={
                  !balance[key]
                    ? ""
                    : "grams" in balance[key]
                      ? balance[key].grams
                      : balance[key].count * gramsPerCoin
                }
                onChange={(e) =>
                  setBalance((old) => ({
                    ...old,
                    [key]: { grams: parseFloat(e.target.value) },
                  }))
                }
              />
            </label>
          </div>
        ))}
      </div>
      <button type={"submit"} disabled={!department || !hasBalance()}>
        Submit
      </button>
    </form>
  );
}
