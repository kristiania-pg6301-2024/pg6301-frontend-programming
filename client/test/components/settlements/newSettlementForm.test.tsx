import { afterEach, describe, expect, it, vitest } from "vitest";
import { cleanup, fireEvent, render } from "@testing-library/react";
import { NewSettlementForm } from "../../../src/components/settlements/newSettlementForm";
import React from "react";

afterEach(cleanup);

function fireChange(element: HTMLElement, value: string) {
  fireEvent.change(element, { target: { value: value } });
}

describe("NewSettlementForm", () => {
  it("displays form", () => {
    const app = render(<NewSettlementForm onNewSettlement={vitest.fn()} />);
    expect(app.baseElement).toMatchSnapshot();
  });
  it("creates new settlement", async () => {
    const onNewSettlement = vitest.fn();
    const app = render(<NewSettlementForm onNewSettlement={onNewSettlement} />);
    const department = "Furniture";
    fireChange(await app.findByLabelText("Department:"), department);
    fireChange(await app.findByLabelText("100kr:"), "4");
    fireEvent.click(await app.findByText("Submit"));
    expect(onNewSettlement).toBeCalledWith({
      department,
      balance: { "100kr": 4 },
    });
  });
  it("prevents submitting form without department", async () => {
    const onNewSettlement = vitest.fn();
    const app = render(<NewSettlementForm onNewSettlement={onNewSettlement} />);
    fireChange(await app.findByLabelText("Department:"), "");
    const submitButton = await app.findByText("Submit");
    expect(submitButton.getAttribute("disabled")).toBeDefined();
    fireEvent.click(submitButton);
    expect(onNewSettlement).not.toBeCalled();
  });
});
