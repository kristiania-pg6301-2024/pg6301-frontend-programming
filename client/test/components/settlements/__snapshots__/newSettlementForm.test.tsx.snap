// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NewSettlementForm > displays form 1`] = `
<body>
  <div>
    <form>
      <h1>
        New Settlement
      </h1>
      <div>
        <label
          for="newSettlementDepartment"
        >
          Department: 
        </label>
        <select
          id="newSettlementDepartment"
        >
          <option />
          <option>
            Furniture
          </option>
          <option>
            Books
          </option>
        </select>
      </div>
      <div>
        <label
          for="newSettlementBalance_50kr"
        >
          50kr
          :
        </label>
        <input
          id="newSettlementBalance_50kr"
          type="number"
          value="0"
        />
      </div>
      <div>
        <label
          for="newSettlementBalance_100kr"
        >
          100kr
          :
        </label>
        <input
          id="newSettlementBalance_100kr"
          type="number"
          value="0"
        />
      </div>
      <div>
        <label
          for="newSettlementBalance_200kr"
        >
          200kr
          :
        </label>
        <input
          id="newSettlementBalance_200kr"
          type="number"
          value="0"
        />
      </div>
      <div>
        <label
          for="newSettlementBalance_500kr"
        >
          500kr
          :
        </label>
        <input
          id="newSettlementBalance_500kr"
          type="number"
          value="0"
        />
      </div>
      <div>
        <label
          for="newSettlementBalance_1000kr"
        >
          1000kr
          :
        </label>
        <input
          id="newSettlementBalance_1000kr"
          type="number"
          value="0"
        />
      </div>
      <div>
        <label
          for="newSettlementBalance_20kr"
        >
          20kr
          :
        </label>
        <input
          id="newSettlementBalance_20kr"
          type="number"
        />
      </div>
      <div>
        <label
          for="newSettlementBalance_10kr"
        >
          10kr
          :
        </label>
        <input
          id="newSettlementBalance_10kr"
          type="number"
        />
      </div>
      <div>
        <label
          for="newSettlementBalance_5kr"
        >
          5kr
          :
        </label>
        <input
          id="newSettlementBalance_5kr"
          type="number"
        />
      </div>
      <div>
        <label
          for="newSettlementBalance_1kr"
        >
          1kr
          :
        </label>
        <input
          id="newSettlementBalance_1kr"
          type="number"
        />
      </div>
      <button
        disabled=""
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</body>
`;
